lib/ZnapZend/ZFS.pm

```diff
@@ -948,45 +1038,169 @@ sub getSnapshotProperties {
+    # zfs snapshot of.
     my $recurse = shift; # May be not passed => undef
 
     # NOTE that some versions of zfs do not inherit values from same-named
     # snapshot of a parent dataset, but only from a "real" dataset higher
     # in the data hierarchy, so this flag may be effectively ignored by OS.
+    #   0 = only local
+    #   1 = local + inherit as defined by zfs
+    #   2 = local + recurse into parent that has same snapname
+    #   3 = local + inherit as defined by zfs + recurse into parent
     my $inherit = shift; # May be not passed => undef
 
+    # Limit the request to `zfs get` to only pick out certain properties
+    # and save time not-processing stuff the caller will ignore?..
+    my $propnames = shift; # May be not passed => undef
+    $propnames = $self->filterPropertyNames($propnames); # Returns a string to pass into `zfs get`
+
+    $self->zLog->debug("=== getSnapshotProperties():\n"
+        . "\trecurse=" . Dumper($recurse)
+        . "\tinherit=" . Dumper($inherit)
+        . "\tsnapshot=" . Dumper($snapshot)
+        . "\tpropnames=" . Dumper($propnames)
+#        . "\tlowmemRecurse=" . $self->lowmemRecurse
+        ) if $self->debug;
+
     if (!defined($recurse)) {
         $recurse = 0;
     }
 
     if (!defined($inherit)) {
         $inherit = 0;
+    } else {
+        if ( !(($inherit >= 0) && ($inherit <= 3)) ) {
+            $self->zLog->warn("getSnapshotProperties(): inherit flag is not a number between 0 and 3");
+            $inherit = 1; # caller DID set something...
+        }
     }
+    my $inhMode = 'local';
+    if ($inherit == 1 || $inherit == 3) { $inhMode .= ',inherited'; }
 
     my %properties;
     my $propertyPrefix = $self->propertyPrefix;
 
-    my @cmd = (@{$self->priv}, qw(zfs get -H));
-    if ($inherit) {
-        push (@cmd, qw(-s), 'local,inherited');
-    } else {
-        push (@cmd, qw(-s local));
-    }
+    my @cmd = (@{$self->priv}, qw(zfs get -H -s), $inhMode);
     if ($self->zfsGetType) {
         push (@cmd, qw(-t snapshot));
     }
     if ($recurse) {
         push (@cmd, qw(-r));
     }
-    push (@cmd, qw(-o), 'property,value', 'all', $snapshot);
+    push (@cmd, qw(-o), 'property,value', $propnames, $snapshot);
 
     print STDERR '# ' . join(' ', @cmd) . "\n" if $self->debug;
-    open my $props, '-|', @cmd or Mojo::Exception->throw('ERROR: could not get zfs properties');
+    open my $props, '-|', @cmd or Mojo::Exception->throw('ERROR: could not get zfs properties of ' . $snapshot);
     while (my $prop = <$props>){
         chomp $prop;
         my ($key, $value) = $prop =~ /^\Q$propertyPrefix\E:(\S+)\s+(.+)$/ or next;
         $properties{$key} = $value;
     }
+    my $numProps = keys %properties;
+
+    if ($self->debug) {
+        if ($numProps > 0) {
+            $self->zLog->debug("=== getSnapshotProperties(): GOT $inhMode (#$inherit) properties of $snapshot : " .Dumper(\%properties) );
+        }
+    }
+
+    if ($inherit == 2 || $inherit == 3) {
```

Comment:
I find such long multi level if/then constructs a bit hard to follow ... how about returning the `\%properties if $inherit < 2` here to loose a level ?

Reply:
I originally thought to make it a bit mask, then settled for a number of magic values with specific meanings (commented above) that are a bit faster to compare than more meaningful strings or further variables - and we pass this codepath a lot when we do :\

Reply:
oh I did not realize that ... please don't :) use a hash with speaking  keys for the features you want enabled ...

Reply:
I think I get the idea but did not find an example to rip off, do you have some handy please?

Reply:
```perl
my $inherit = {
    all => 1,
    greens => 0,
    reds => 1
};


if ($inherit->{all} or $inherit->{greens}){ .... 

}
```

Reply:
Ah, just so? I thought there was something computationally cheaper, like C enums. Thanks, will try.

Reply:
I suppose my fix is not what you wanted (ended up like an enum), but cursorily seems to work and steps away from arcane numbers to human names for them, but still uses the number underneath. Might revise later to have it as a flag map per your example, but I think it would be a refactor irrelevant to functionality at the moment :)